annotated_test.go

```diff
@@ -0,0 +1,39 @@
+package fx_test
+
+import (
+	"testing"
+	"go.uber.org/fx/fxtest"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/fx"
+)
+
+func TestAnnotated(t *testing.T) {
+	type a struct {
+		name string
+	}
+
+	type in struct {
+		fx.In
+
+		A *a `name:"foo"`
+	}
+	newA := func() *a {
+		return &a{name: "foo"}
+	}
+
+	t.Run("Provided", func(t *testing.T) {
+		app := fxtest.New(t,
+			fx.Provide(
+				fx.Annotated{
+					Name: "foo",
+					Fn:   newA,
+				},
+			),
+			fx.Invoke(func(in in) {
+				assert.NotNil(t, in.A, "expected in.A to be injected")
+				assert.Equal(t, "foo", in.A.name, "expected to get a type 'a' of name 'foo'")
```

Comment:
Instead of putting assertions in the `Invoke`, consider moving them out with
the use of an [`fx.Populate`](https://godoc.org/go.uber.org/fx#Populate).

    var result struct {
        fx.In
        
        A *a `name:"foo"`
    }
    
    app := fxtest.New(t,
        ...,
        fx.Populate(&result),
    )
    defer app.RequireStart().RequireStop()
    
    assert.[..]
    
Otherwise, if we have a bug that neglects to call that `fx.Invoke`, the
assertion will be ignored and the test will still pass.